### DISCUSSION TITLE
OSC message to MIDI note/velocity

### DISCUSSION THREAD

**Discussion by coachz1** [helgoboss/helgobox#890] [Q&A] [answered]
I got the fine Realearn seeing my OSC inputs when I press pads on OSC Controller in Android and I got it converting it to a MIDI note and velocity.   

I used the following and got midi working but I a long latency even though my Reaper is very low latency at 128 buffer and all other instruments are fast response.  Any ideas please ?
F3		90 41 64
F#3		90 42 64

![2023-10-20 09_40_48-ztemp  modified  - REAPER v7 001+dev1018a - Licensed for personal_small business](https://github.com/helgoboss/realearn/assets/53879358/75fdb9be-ba72-470f-9bf8-1dd4301bd58d)

![2023-10-20 09_38_05-sTabLauncher](https://github.com/helgoboss/realearn/assets/53879358/4601fc69-a66f-4cbd-a679-f2ee650fc866)

**Comment by Leamucho** [comment]
Hey, @coachz1 you don't have anything set at the OUTPUT on the most general side of Realearn. Then, defining the Output of the Target, you left it with FX Output, so you must be sure to put the synth just after the Realearn instance, and if you have it on monitorfx... 
So, put it on Feedbackoutput, and set a virtual Midi thread on reaper...  that same is what you select  as OUTPUT at the first part... then you put it as input for the track where you put your synth...
Try and adhoc connection also for the tablet-computer connection, as wifi  might get busy with other stuff...
take care

**Comment by helgoboss** [comment, marked as answer]
Sorry for the late response. GitHub discussions was not much on my radar so far. But now that it is becoming the main way to ask question, I better answer this one ;)

The simple answer: What you want to do is currently not possible in ReaLearn with low latency.

The way OSC messages are processed currently:

1. ReaLearn gets aware of them quite quickly, not much latency. It depends on your network really.
2. ReaLearn processes the message on the next main thread cycle. And this is the problem. This can take up to 30ms.

For most parameter control scenarios, this is not an issue. But for playing an instrument, it is.

If you need that, please create a feature request (in the Ideas category of GitHub Discussion). Maybe there's one already, not even sure.
